File: AngularFrontEnd/src/app/controller-calls/booking.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { BookingResponse } from '../pojos/BookingResponse';

@Injectable({
  providedIn: 'root'
})
export class BookingService {

  constructor(private http:HttpClient) { }

  public getBookingById(bookingId: String): Observable<BookingResponse> {
    return this.http.get<BookingResponse>("http://localhost:8080/v1/bookings/"+bookingId);
  }

  public getBookingsByUserId(userId: String): Observable<BookingResponse[]> {
    return this.http.get<BookingResponse[]>("http://localhost:8080/v1/bookings/user/"+userId);
  }

  public getBookingsByDriverId(driverId: String): Observable<BookingResponse[]>  {
    return this.http.get<BookingResponse[]>("http://localhost:8080/v1/bookings/driver/"+driverId);
  }

  public getAllBookings(): Observable<BookingResponse[]>  {
    return this.http.get<BookingResponse[]>("http://localhost:8080/v1/bookings/adminBookings");
  }

  public createBooking(booking: any): Observable<BookingResponse> {
    return this.http.post<BookingResponse>("http://localhost:8080/v1/bookings/registerBooking", booking, {responseType: 'json'});
  }

  public updateBooking(bookingId: String, updatedBooking: any): Observable<BookingResponse> {
    return this.http.put<BookingResponse>("http://localhost:8080/v1/bookings/update/"+bookingId, updatedBooking);
  }

  public deleteBooking(bookingId: String): Observable<BookingResponse> {
    return this.http.delete<BookingResponse>("http://localhost:8080/v1/bookings/"+bookingId);
  }
}

File: SpringBootProject/src/main/java/com/caltech/service/BookingService.java
package com.caltech.service;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.caltech.pojo.Booking;
import com.caltech.repository.BookingRepository;

@Service
public class BookingService {

	@Autowired
	private BookingRepository bookingRepository;
	
	@Autowired
	private FareCalculatorService fairCalculatorService;
	
    public Booking getBookingById(UUID bookingId) {
        Optional<Booking> optionalBooking = bookingRepository.findById(bookingId);
        return optionalBooking.orElse(null);
    }
    
    public List<Booking> getBookingsByUserId(UUID userId) {
        return bookingRepository.findByUserId(userId);
    }
    
    public List<Booking> getBookingsByDriverId(UUID driverId) {
        return bookingRepository.findByDriverId(driverId);
    }
    
    public List<Booking> getAllBookings() {
        return bookingRepository.findAll();
    }
    
    public Booking createBooking(Booking booking) {
    	booking.setTravelFare(fairCalculatorService.calculateFare(booking));
        return bookingRepository.save(booking);
    }
    
    public Booking updateBooking(UUID bookingId, Booking updatedBooking) {
        Optional<Booking> optionalBooking = bookingRepository.findById(bookingId);
        if (optionalBooking.isPresent()) {
            Booking existingBooking = optionalBooking.get();
            // Update fields based on your requirements
            existingBooking.setSource(updatedBooking.getSource());
            existingBooking.setDestination(updatedBooking.getDestination());
            existingBooking.setBookingTime(updatedBooking.getBookingTime());
            // per update will result in an additional $3 surcharge
            existingBooking.setTravelFare(existingBooking.getTravelFare() + 3.00);
            return bookingRepository.save(existingBooking);
        }
        return null;
    }
    
    public void deleteBooking(UUID bookingId) {
        Optional<Booking> optionalBooking = bookingRepository.findById(bookingId);

        if (optionalBooking.isPresent()) {
            Booking booking = optionalBooking.get();

            // Disassociate from User
            if (booking.getUser() != null) {
                List<Booking> userBookings = booking.getUser().getBookings();
                if (userBookings != null) {
                    userBookings.remove(booking);
                }
                booking.setUser(null);
            }

            // Disassociate from Driver
            if (booking.getDriver() != null) {
                List<Booking> driverBookings = booking.getDriver().getBookings();
                if (driverBookings != null) {
                    driverBookings.remove(booking);
                }
                booking.setDriver(null);
            }

            // Save the modified entities
            bookingRepository.save(booking);

            // Delete the Booking
            bookingRepository.delete(booking);
        }
	}
}
